src/dyngen.cpp

```diff
@@ -14,199 +14,386 @@ static CodeLibrary *gLibrary = nullptr;
-      while (unit!=nullptr) {
-        if (unit->dynGenUnit == this) {
-          if (previousUnit != nullptr) {
-            if (unit->next != nullptr) {
-              previousUnit->next = unit->next;
-            } else {
-              previousUnit->next = nullptr;
-            }
-          } else {
-            node->dynGenNodes = nullptr;
-          }
-          ft->fRTFree(mWorld, static_cast<void*>(unit));
-          break;
-        }
-        previousUnit = unit;
-        unit = unit->next;
-      }
-    };
-    node = node->next;
+  mStub->mObject = nullptr;
+  mStub->mRefCount -= 1;
+  if (mStub->mRefCount==0) {
+    RTFree(mWorld, mStub);
+  }
+
+  // in case we have not found a matching code the vm was never initialized
+  // so we also do not have access to a code library or a vm we would have
+  // to clean up, so we bail out early.
+  if (mCodeLibrary == nullptr) {
+    return;
+  }
+  // readjust the head of the linked list of the code library if necessary
+  if (mCodeLibrary->dynGen == this) {
+    mCodeLibrary->dynGen = mNextDynGen;
+  }
+
+  // remove ourselves from the linked list
+  if (mPrevDynGen != nullptr) {
+    mPrevDynGen->mNextDynGen = mNextDynGen;
   }
+  if (mNextDynGen != nullptr) {
+    mNextDynGen->mPrevDynGen = mPrevDynGen;
+  }
+
+  // free the vm in RT context through async command
+  ft->fDoAsynchronousCommand(
+    mWorld,
+    nullptr,
+    nullptr,
+    static_cast<void*>(mVm),
+    deleteVmOnSynthDestruction,
+    nullptr,
+    nullptr,
+    doNothing,
+    0,
+    nullptr
+  );
 }
 
 void DynGen::next(int numSamples) {
-  if (!vm.mReady.load()) {
+  if (mVm == nullptr) {
     for (int i = 0; i < mNumOutputs; i++) {
       Clear(numSamples, mOutBuf[i]);
     }
   } else {
     // skip first 2 channels since those are not signals
-    vm.process(mInBuf + 2, mOutBuf, numSamples);
+    mVm->process(mInBuf + 2, mOutBuf, numSamples);
   }
 }
 
-void DynGen::updateCode(const std::string &code) {
-  vm.mReady.store(false, std::memory_order_release);
-  // init sets the mReady variable when finished/successful
-  vm.init(code);
-}
-
-
 // ************
 // Library code
 // ************
 
-// this runs in stage 2 (NRT) and enters the content of the file to
-// the library by traversing the library which is a linked list.
-// If the hash ID already exists the code gets updated and all running
-// instances should be updated.
-bool enterFileToDynGenLibrary(World *world, void *raw_callback) {
-  auto entry = static_cast<NewDynGenLibraryEntry*>(raw_callback);
+// this runs in stage 2 (NRT) and loads the content of the file
+// which gets passed to stage 3 (RT)
+bool loadFileToDynGenLibrary(World *world, void *rawCallbackData) {
+  auto entry = static_cast<NewDynGenLibraryEntry*>(rawCallbackData);
+
+  auto codeFile = std::ifstream(entry->codePath, std::ios::binary);
+  if (!codeFile.is_open()) {
+    std::cerr << "Could not open DynGen file at " << entry->codePath << std::endl;
```

Comment:
I think you should use `Print` instead of `std::cerr`. If you start the message with `ERROR:`, it will be displayed as an error message in the IDE.